File: src/main/java/de/donuz/plus4/CpuState.java
/**
 * plus4
 * Copyright (C) 2009 Florian Wolff  (florian@donuz.de)
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 *
 * $Id: CpuState.java,v 1.8 2009/06/14 15:49:49 florian Exp $
 */
package de.donuz.plus4;

public class CpuState
{

    public int pc;
    public short acc;
    public short xr;
    public short yr;
    public int sp;
    public boolean c;
    public boolean z;
    public boolean n;
    public boolean b;
    public boolean v;
    public boolean i;
    public boolean d;
    
    
    public void incrPc(int inc)
    {
        pc += inc;
        pc &= 0xffff;
    }


    public void incrSp()
    {
        sp ++;
        sp &= 0xff;
    }


    public void decrSp()
    {
        sp --;
        if (sp < 0) sp = 0xff;
    }


    void NegFlag(short by)
    { 
        n = ((by & 0x80) > 0); 
    }

    void ZeroFlag(short by)
    {
        z = (by == 0);
    }

    public void DoADC(short data)
    {
        short oldAcc = acc;

        if (d) {
            // Decimal Mode (BCD)
            final int carryIn = c ? 1 : 0;

            // Add low nibbles
            int low = (acc & 0x0F) + (data & 0x0F) + carryIn;
            if (low > 0x09) {
                low += 0x06;
            }

            // Add high nibbles
            int high = (acc >> 4) + (data >> 4) + (low > 0x0F ? 1 : 0);

            // Set flags based on binary result for compatibility
            final int binResult = acc + data + carryIn;
            z = ((binResult & 0xFF) == 0);
            n = ((binResult & 0x80) != 0);
            v = (((~(oldAcc ^ data)) & (oldAcc ^ binResult) & 0x80) != 0);

            if (high > 0x09) {
                high += 0x06;
            }

            c = (high > 0x0F);
            acc = toByte(((high << 4) | (low & 0x0F)) & 0xFF);
        } else {
            // Binary Mode:
            final int raw = data + acc + (c ? 1 : 0);
            acc = toByte(raw);
            c = (raw > 0xff);

            NegFlag(acc);
            ZeroFlag(acc);

            // Overflow flag: set if sign bit is incorrect
            v = (((~(oldAcc ^ data)) & (oldAcc ^ acc) & 0x80) != 0);
        }
    }
    
    public void DoAND(short second)
    {
        acc &= second;
        NegFlag(acc);
        ZeroFlag(acc);
    }
    
    public short DoASL(short data)
    {        
        c = ((data & 0x80) > 0);
        data <<= 1;
        data &= 0xff;
        ZeroFlag(data);
        NegFlag(data);
        
        return data;
    }

    public void DoBIT(short data)
    {
        v = ((data & 64) > 0);
        n = ((data & 128) > 0);
        z = ((data & acc) == 0);
    }

    public void DoCMP(short reg, short data)
    {
        c = (reg >= data);
        z = (data == reg);
        n = (toByte(reg - data) > 0x7f);
    }

    public void DoEOR(short value)
    {
        acc ^= value;
        acc &= 0xff;
        NegFlag(acc);
        ZeroFlag(acc);
    }

    public short DoLSR(short data)
    {
        c = ((data & 0x01) > 0);
        data >>= 1;
        ZeroFlag(data);
        n = false;

        return data;
    }

    public void DoORA(short value)
    {
        acc |= value;
        NegFlag(acc);
        ZeroFlag(acc);
    }

    public short DoROL(short data)
    {
        final boolean newC = ((data & 0x80) > 0);
        data <<= 1;
        data &= 0xff;
        if (c) {
            data |= 1;
        }
        c = newC;
        NegFlag(data);
        ZeroFlag(data);
        
        return data;
    }

    public short DoROR(short data)
    {
        final boolean newC = ((data & 1) > 0);
        data >>= 1;
        if (c) { 
            data |= 0x80;
        }
        c = newC;
        NegFlag(data);
        ZeroFlag(data);

        return data;
    }

    public void DoSBC(short data)
    {
        final short oldAcc = acc;

        if (d) {
            int tmpDec = (acc & 0x0f) - (data & 0x0f) - (c ? 0 : 1);

            if ((tmpDec & 0x10) > 0) {                                       
                tmpDec = ((tmpDec - 6) & 0xf) | ((acc & 0xf0) - (data & 0xf0) - 0x10);
            } else {                                                 
                tmpDec = (tmpDec & 0xf) | ((acc & 0xf0) - (data & 0xf0));
            }
            if ((tmpDec & 0x100) > 0) {
                tmpDec -= 0x60;                                                 
            }
            
            acc = toByte(tmpDec);
            c = (tmpDec < 0x100);

            NegFlag(acc);
            ZeroFlag(acc);

            v = false;
            if (!n && (acc & 0x80)>0  && (data & 0x80)>0  && (oldAcc & 0x80)==0) v = true;
            if (!n && (acc & 0x80)==0 && (data & 0x80)==0 && (oldAcc & 0x80)>0 ) v = true;
            if ( n && (acc & 0x80)>0  && (data & 0x80)>0  && (oldAcc & 0x80)==0) v = true;
        } else {
            final int raw = acc - data - (c ? 0 : 1);

            acc = toByte(raw);
            c = (raw >= 0);  // Carry set if no borrow occurred

            NegFlag(acc);
            ZeroFlag(acc);

            // Overflow flag: set if sign bit is incorrect
            v = (((oldAcc ^ data) & (oldAcc ^ acc) & 0x80) != 0);
        }
    }

    public void dey()
    {
        yr--;
        if (yr < 0) {
            yr = 0xff;
        }
        NegFlag(yr);
        ZeroFlag(yr);
    }


    public void iny()
    {
        yr = (short) ((yr + 1) & 0xff);
        NegFlag(yr);
        ZeroFlag(yr);
    }


    public void dex()
    {
        xr--;
        if (xr < 0) {
            xr = 0xff;
        }
        NegFlag(xr);
        ZeroFlag(xr);
    }


    public void inx()
    {
        xr = (short) ((xr + 1) & 0xff);
        NegFlag(xr);
        ZeroFlag(xr);
    }


    public void txa()
    {
        acc = xr;
        NegFlag(acc);
        ZeroFlag(acc);
    }


    public void tax()
    {
        xr = acc;
        NegFlag(xr);
        ZeroFlag(xr);
    }


    public void tya()
    {
        acc = yr;
        NegFlag(acc);
        ZeroFlag(acc);
    }


    public void tay()
    {
        yr = acc;
        NegFlag(yr);
        ZeroFlag(yr);
    }


    public void tsx()
    {
        xr = (short) sp;
        NegFlag(xr);
        ZeroFlag(xr);
    }


    public void txs()
    {
        sp = xr;
    }

    
    public static short toByte(int b)
    {
        if (b > 0xff) {
            return (short) (b & 0xff);
        } else {
            while (b < 0) {
                b += 0x100;
            }
        
            return (short) b;
        }
    }
}

File: src/cpu_state.rs
/// CPU State for 6510 processor
/// Copyright (C) 2009 Florian Wolff (florian@donuz.de)
/// Rust port 2025
///
/// This program is free software; you can redistribute it and/or
/// modify it under the terms of the GNU General Public License
/// as published by the Free Software Foundation; either version 2
/// of the License, or (at your option) any later version.

#[derive(Debug, Clone)]
pub struct CpuState {
    // Registers
    pub pc: u16,      // Program Counter
    pub acc: u8,      // Accumulator
    pub xr: u8,       // X Register
    pub yr: u8,       // Y Register
    pub sp: u8,       // Stack Pointer

    // Status Flags
    pub c: bool,      // Carry
    pub z: bool,      // Zero
    pub n: bool,      // Negative
    pub b: bool,      // Break
    pub v: bool,      // Overflow
    pub i: bool,      // Interrupt Disable
    pub d: bool,      // Decimal Mode
}

impl Default for CpuState {
    fn default() -> Self {
        Self {
            pc: 0,
            acc: 0,
            xr: 0,
            yr: 0,
            sp: 0xFF,
            c: false,
            z: false,
            n: false,
            b: false,
            v: false,
            i: false,
            d: false,
        }
    }
}

impl CpuState {
    pub fn new() -> Self {
        Self::default()
    }

    // Program Counter operations
    pub fn incr_pc(&mut self, inc: u16) {
        self.pc = self.pc.wrapping_add(inc) & 0xFFFF;
    }

    // Stack Pointer operations
    pub fn incr_sp(&mut self) {
        self.sp = self.sp.wrapping_add(1);
    }

    pub fn decr_sp(&mut self) {
        self.sp = self.sp.wrapping_sub(1);
    }

    // Flag operations
    pub fn neg_flag(&mut self, byte: u8) {
        self.n = (byte & 0x80) != 0;
    }

    pub fn zero_flag(&mut self, byte: u8) {
        self.z = byte == 0;
    }

    // ADC - Add with Carry
    pub fn do_adc(&mut self, data: u8) {
        let old_acc = self.acc;

        if self.d {
            // Decimal mode not implemented yet
            panic!("Decimal mode ADC not implemented");
        } else {
            // Binary mode
            let raw = data as u16 + self.acc as u16 + if self.c { 1 } else { 0 };
            self.acc = raw as u8;
            self.c = raw > 0xFF;

            self.neg_flag(self.acc);
            self.zero_flag(self.acc);

            self.v = false;
            if self.n && (self.acc & 0x80) != 0 && (data & 0x80) == 0 && (old_acc & 0x80) == 0 {
                self.v = true;
            }
            if self.n && (self.acc & 0x80) == 0 && (data & 0x80) != 0 && (old_acc & 0x80) != 0 {
                self.v = true;
            }
            if !self.n && (self.acc & 0x80) == 0 && (data & 0x80) != 0 && (old_acc & 0x80) != 0 {
                self.v = true;
            }
        }
    }

    // AND - Logical AND
    pub fn do_and(&mut self, second: u8) {
        self.acc &= second;
        self.neg_flag(self.acc);
        self.zero_flag(self.acc);
    }

    // ASL - Arithmetic Shift Left
    pub fn do_asl(&mut self, data: u8) -> u8 {
        self.c = (data & 0x80) != 0;
        let result = data << 1;
        self.zero_flag(result);
        self.neg_flag(result);
        result
    }

    // BIT - Bit Test
    pub fn do_bit(&mut self, data: u8) {
        self.v = (data & 64) != 0;
        self.n = (data & 128) != 0;
        self.z = (data & self.acc) == 0;
    }

    // CMP - Compare
    pub fn do_cmp(&mut self, reg: u8, data: u8) {
        self.c = reg >= data;
        self.z = data == reg;
        self.n = reg.wrapping_sub(data) > 0x7F;
    }

    // EOR - Exclusive OR
    pub fn do_eor(&mut self, value: u8) {
        self.acc ^= value;
        self.neg_flag(self.acc);
        self.zero_flag(self.acc);
    }

    // LSR - Logical Shift Right
    pub fn do_lsr(&mut self, data: u8) -> u8 {
        self.c = (data & 0x01) != 0;
        let result = data >> 1;
        self.zero_flag(result);
        self.n = false;
        result
    }

    // ORA - Logical OR
    pub fn do_ora(&mut self, value: u8) {
        self.acc |= value;
        self.neg_flag(self.acc);
        self.zero_flag(self.acc);
    }

    // ROL - Rotate Left
    pub fn do_rol(&mut self, data: u8) -> u8 {
        let new_c = (data & 0x80) != 0;
        let mut result = data << 1;
        if self.c {
            result |= 1;
        }
        self.c = new_c;
        self.neg_flag(result);
        self.zero_flag(result);
        result
    }

    // ROR - Rotate Right
    pub fn do_ror(&mut self, data: u8) -> u8 {
        let new_c = (data & 1) != 0;
        let mut result = data >> 1;
        if self.c {
            result |= 0x80;
        }
        self.c = new_c;
        self.neg_flag(result);
        self.zero_flag(result);
        result
    }

    // SBC - Subtract with Carry
    pub fn do_sbc(&mut self, data: u8) {
        let old_acc = self.acc;

        if self.d {
            // Decimal mode
            let mut tmp_dec = (self.acc as i32 & 0x0F) - (data as i32 & 0x0F) - if self.c { 0 } else { 1 };

            if (tmp_dec & 0x10) != 0 {
                tmp_dec = ((tmp_dec - 6) & 0xF) | ((self.acc as i32 & 0xF0) - (data as i32 & 0xF0) - 0x10);
            } else {
                tmp_dec = (tmp_dec & 0xF) | ((self.acc as i32 & 0xF0) - (data as i32 & 0xF0));
            }
            if (tmp_dec & 0x100) != 0 {
                tmp_dec -= 0x60;
            }

            self.acc = tmp_dec as u8;
            self.c = tmp_dec < 0x100;

            self.neg_flag(self.acc);
            self.zero_flag(self.acc);

            self.v = false;
            if !self.n && (self.acc & 0x80) != 0 && (data & 0x80) != 0 && (old_acc & 0x80) == 0 {
                self.v = true;
            }
            if !self.n && (self.acc & 0x80) == 0 && (data & 0x80) == 0 && (old_acc & 0x80) != 0 {
                self.v = true;
            }
            if self.n && (self.acc & 0x80) != 0 && (data & 0x80) != 0 && (old_acc & 0x80) == 0 {
                self.v = true;
            }
        } else {
            // Binary mode
            let raw = self.acc as i32 - data as i32 - if self.c { 0 } else { 1 };

            self.acc = raw as u8;
            self.c = raw < 0x100;

            self.neg_flag(self.acc);
            self.zero_flag(self.acc);

            self.v = false;
            if !self.n && (self.acc & 0x80) != 0 && (data & 0x80) != 0 && (old_acc & 0x80) == 0 {
                self.v = true;
            }
            if !self.n && (self.acc & 0x80) == 0 && (data & 0x80) == 0 && (old_acc & 0x80) != 0 {
                self.v = true;
            }
            if self.n && (self.acc & 0x80) != 0 && (data & 0x80) != 0 && (old_acc & 0x80) == 0 {
                self.v = true;
            }
        }
    }

    // Register operations
    pub fn dey(&mut self) {
        self.yr = self.yr.wrapping_sub(1);
        self.neg_flag(self.yr);
        self.zero_flag(self.yr);
    }

    pub fn iny(&mut self) {
        self.yr = self.yr.wrapping_add(1);
        self.neg_flag(self.yr);
        self.zero_flag(self.yr);
    }

    pub fn dex(&mut self) {
        self.xr = self.xr.wrapping_sub(1);
        self.neg_flag(self.xr);
        self.zero_flag(self.xr);
    }

    pub fn inx(&mut self) {
        self.xr = self.xr.wrapping_add(1);
        self.neg_flag(self.xr);
        self.zero_flag(self.xr);
    }

    // Transfer operations
    pub fn txa(&mut self) {
        self.acc = self.xr;
        self.neg_flag(self.acc);
        self.zero_flag(self.acc);
    }

    pub fn tax(&mut self) {
        self.xr = self.acc;
        self.neg_flag(self.xr);
        self.zero_flag(self.xr);
    }

    pub fn tya(&mut self) {
        self.acc = self.yr;
        self.neg_flag(self.acc);
        self.zero_flag(self.acc);
    }

    pub fn tay(&mut self) {
        self.yr = self.acc;
        self.neg_flag(self.yr);
        self.zero_flag(self.yr);
    }

    pub fn tsx(&mut self) {
        self.xr = self.sp;
        self.neg_flag(self.xr);
        self.zero_flag(self.xr);
    }

    pub fn txs(&mut self) {
        self.sp = self.xr;
    }
}
